cli/src/main/java/org/eclipse/hono/cli/CommandAndControlClient.java

```diff
@@ -0,0 +1,87 @@
+/*******************************************************************************
+ * Copyright (c) 2016, 2018 Contributors to the Eclipse Foundation
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information regarding copyright ownership.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ *******************************************************************************/
+
+package org.eclipse.hono.cli;
+
+import java.net.HttpURLConnection;
+import java.util.concurrent.CompletableFuture;
+
+import org.apache.qpid.proton.message.Message;
+import org.eclipse.hono.util.CommandConstants;
+import org.eclipse.hono.util.MessageHelper;
+import org.springframework.context.annotation.Profile;
+import org.springframework.stereotype.Component;
+
+import io.vertx.core.Future;
+import io.vertx.proton.ProtonDelivery;
+import io.vertx.proton.ProtonHelper;
+import io.vertx.proton.ProtonMessageHandler;
+import io.vertx.proton.ProtonQoS;
+import io.vertx.proton.ProtonReceiver;
+
+/**
+ * A client for accessing Hono's Command and Control APIs to
+ * receive and send command response messages via the AMQP adapter.
+ */
+@Component
+@Profile("amqp-command")
+public class CommandAndControlClient extends AmqpCliClient {
+
+    /**
+     * Connects to the AMQP adapter and open a command receiver link for receiving commands from the adapter. Once a
+     * command request message is received, this method automatically sends a command response back to the adapter.
+     *
+     * @param messageTracker The future to notify when the message is sent. The future is completed with the delivery
+     *            upon success or complete the future with an exception.
+     */
+    @Override
+    protected void sendMessage(final CompletableFuture<ProtonDelivery> messageTracker) {
+        connectToAdapter()
+        .compose(con -> {
+            adapterConnection = con;
+            return createSender();
+        }).map(sender -> subscribeToCommands((d, m) -> {
+            LOG.info("Received Command Message : {}", m.getSubject());
+            final Message commandResponse = ProtonHelper.message(m.getReplyTo(), "OK: " + m.getSubject());
+            commandResponse.setCorrelationId(m.getCorrelationId());
+            MessageHelper.addProperty(commandResponse, MessageHelper.APP_PROPERTY_STATUS, HttpURLConnection.HTTP_OK);
+            commandResponse.setContentType(m.getContentType());
+            sender.send(commandResponse, delivery -> messageTracker.complete(delivery));
```

Comment:
I do not see any relation between sending a command response and completing this future ...
The command consumer is supposed to run until the user hits ctrl-c and should reply to all commands that arrive, right?

Reply:
> The command consumer is supposed to run until the user hits ctrl-c and should reply to all commands that arrive, right?

what i had in mind was for the user to start a command consumer to receive -> send command response -> then exit

i will change this so the consumer runs until explicitly closed

Reply:
> i will change this so the consumer runs until explicitly closed

don't you think that this is more practical?

Reply:
yes, way more practical and intuitive :-)